Shoko.Server/ImageDownload/ImageDownloadRequest.cs

```diff
@@ -1,17 +1,126 @@
+            AniDB_Seiyuu creator => string.Format(ImageServerUrl, creator.PicName),
+            AnimeStaff staff => string.Format(ImageServerUrl, staff.ImagePath),
+            MovieDB_Fanart movieFanart => string.Format(Constants.URLS.MovieDB_Images, movieFanart.URL),
+            MovieDB_Poster moviePoster => string.Format(Constants.URLS.MovieDB_Images, moviePoster.URL),
+            SVR_AniDB_Anime anime => string.Format(ImageServerUrl, anime.Picname),
+            TvDB_Episode ep => string.Format(Constants.URLS.TvDB_Episode_Images, ep.Filename),
+            TvDB_ImageFanart fanart => string.Format(Constants.URLS.TvDB_Images, fanart.BannerPath),
+            TvDB_ImagePoster poster => string.Format(Constants.URLS.TvDB_Images, poster.BannerPath),
+            TvDB_ImageWideBanner wideBanner => string.Format(Constants.URLS.TvDB_Images, wideBanner.BannerPath),
+            _ => string.Empty
+        };
+
+    public ImageDownloadRequest(object data, string? imageServerUrl = null)
     {
-        ImageType = imageType;
         ImageData = data;
-        ForceDownload = forceDownload;
+        ImageServerUrl = imageServerUrl ?? "";
+    }
+
+    public string? DownloadNow(int maxRetries = 5)
+        => RecursivelyRetryDownload(0, maxRetries);
+
+    private string? RecursivelyRetryDownload(int count, int maxRetries)
+    {
+        // Abort if the download url or final destination is not available.
+        if (string.IsNullOrEmpty(DownloadUrl) || string.IsNullOrEmpty(FilePath))
+            return null;
+
+        var tempPath = Path.Combine(ImageUtils.GetImagesTempFolder(), Path.GetFileName(FilePath));
+        try
+        {
+            // Ignore all certificate failures.
+            ServicePointManager.Expect100Continue = true;
+            ServicePointManager.SecurityProtocol = SecurityProtocolType.Tls12;
+            ServicePointManager.ServerCertificateValidationCallback = delegate { return true; };
+
+            // Download the image.
+            using (var client = new HttpClient())
+            {
+                // Download the image data.
+                client.DefaultRequestHeaders.Add("user-agent", "JMM");
+                var bytes = client.GetByteArrayAsync(DownloadUrl)
+                    .ConfigureAwait(false)
+                    .GetAwaiter()
+                    .GetResult();
+                if (bytes.Length < 4)
+                    throw new WebException(
+                        "The image download stream returned less than 4 bytes (a valid image has 2-4 bytes in the header)");
+
+                // Get the file extension.
+                var extension = Misc.GetImageFormat(bytes) switch
+                {
+                    string imageFormat => "." + imageFormat,
+                    _ => throw new WebException("The image download stream returned an invalid image"),
+                };
+
+                // Change the file extension.
+                tempPath = Path.ChangeExtension(tempPath, extension);
```

Comment:
I have no idea why we change the file extension _on the temporary file path_ while also **not** changing it on the final file path. To be clear, this was the behavior from before this PR, and I didn't change it to this, only copy it. 

It should probably be removed as it serves no purpose (in my eyes at least :shrug:). Checking the image format should still be done — we only need to check if it's not `null` (now after the last two commits) to check if it's a valid image — but changing the extension on the temp. file is unneeded.

Alternately we can just keep it. I won't make the final call. Either way is fine. :upside_down_face:

Reply:
I agree, it should be done on the final path

Reply:
Will try to change it (so it also updates the database entry)

Reply:
Tl;dr: that's hard, since we're not storing every path in the database, and not all paths are stored the same way.